File: backend/routers/blog.py
import re
import time
import hashlib
import asyncio
import email.utils
from typing import List, Optional

import httpx
import feedparser
from fastapi import APIRouter, HTTPException

from models.schemas import BlogPost

router = APIRouter()

MEDIUM_RSS_URL = "https://medium.com/feed/@paragagnihotri"
CACHE_TTL      = 3600  # seconds (1 hour)

_cache: dict = {"posts": [], "expires_at": 0.0}


def _extract_thumbnail(entry) -> Optional[str]:
    """Return the best thumbnail URL for a feed entry."""
    # feedparser sometimes parses media:thumbnail or media:content
    thumbnails = entry.get("media_thumbnail") or entry.get("media_content") or []
    if thumbnails:
        return thumbnails[0].get("url")

    # Fall back: first <img src="..."> in the full content HTML
    content_html = ""
    if entry.get("content"):
        content_html = entry["content"][0].get("value", "")
    elif entry.get("summary"):
        content_html = entry["summary"]

    match = re.search(r'<img[^>]+src=["\']([^"\']+)["\']', content_html)
    if match:
        return match.group(1)

    return None


def _parse_date(entry) -> str:
    """Return an ISO-8601 date string (YYYY-MM-DD)."""
    raw = entry.get("published") or entry.get("updated") or ""
    try:
        parsed = email.utils.parsedate_to_datetime(raw)
        return parsed.date().isoformat()
    except Exception:
        return raw[:10] if len(raw) >= 10 else ""


def _slug_from_url(url: str) -> str:
    """Extract the slug from a Medium article URL."""
    # e.g. https://medium.com/@user/some-title-abc123 → some-title-abc123
    path = url.split("?")[0].rstrip("/")
    return path.split("/")[-1]


def _entry_to_post(entry) -> BlogPost:
    url     = entry.get("link", "")
    slug    = _slug_from_url(url) or hashlib.md5(url.encode()).hexdigest()[:8]
    post_id = hashlib.md5(url.encode()).hexdigest()[:12]

    tags: List[str] = [
        tag.get("term", "") for tag in entry.get("tags", []) if tag.get("term")
    ]

    # Medium puts the subtitle / first paragraph in summary (HTML)
    raw_summary = entry.get("summary", "")
    description = re.sub(r"<[^>]+>", "", raw_summary).strip()[:400]

    return BlogPost(
        id          = post_id,
        title       = entry.get("title", "Untitled"),
        slug        = slug,
        url         = url,
        thumbnail   = _extract_thumbnail(entry),
        date        = _parse_date(entry),
        description = description,
        tags        = tags,
    )


async def _fetch_posts() -> List[BlogPost]:
    """Fetch and parse the Medium RSS feed asynchronously."""
    async with httpx.AsyncClient(timeout=10) as client:
        resp = await client.get(MEDIUM_RSS_URL, follow_redirects=True)
        resp.raise_for_status()
        raw_xml = resp.text

    # feedparser is synchronous — run it off the event loop
    feed = await asyncio.to_thread(feedparser.parse, raw_xml)

    posts = [_entry_to_post(e) for e in feed.entries]
    posts.sort(key=lambda p: p.date, reverse=True)
    return posts


async def _get_cached_posts() -> List[BlogPost]:
    now = time.monotonic()
    if now < _cache["expires_at"] and _cache["posts"]:
        return _cache["posts"]

    posts = await _fetch_posts()
    _cache["posts"]      = posts
    _cache["expires_at"] = now + CACHE_TTL
    return posts


@router.get("/blog", response_model=List[BlogPost])
async def get_blog_posts() -> List[BlogPost]:
    try:
        return await _get_cached_posts()
    except httpx.HTTPError as exc:
        raise HTTPException(status_code=502, detail=f"Failed to fetch Medium feed: {exc}")
    except Exception as exc:
        raise HTTPException(status_code=500, detail=str(exc))


@router.get("/blog/{slug}", response_model=BlogPost)
async def get_blog_post(slug: str) -> BlogPost:
    posts = await get_blog_posts()
    for post in posts:
        if post.slug == slug:
            return post
    raise HTTPException(status_code=404, detail="Blog post not found")

File: frontend/src/lib/blog.ts
import type { BlogPost } from "@/types";
import crypto from "crypto";

const MEDIUM_RSS_URL = "https://medium.com/feed/@paragagnihotri";
const CACHE_TTL = 3600; // 1 hour in seconds

let cache: {
  posts: BlogPost[];
  expiresAt: number;
} = {
  posts: [],
  expiresAt: 0,
};

function extractThumbnail(itemXml: string): string | undefined {
  // Try to find media:thumbnail
  const mediaThumbnailMatch = itemXml.match(/media:thumbnail[^>]*url=["']([^"']+)["']/);
  if (mediaThumbnailMatch) {
    return mediaThumbnailMatch[1];
  }

  // Try to find first <img src="..."> in content
  const imgMatch = itemXml.match(/<img[^>]+src=["']([^"']+)["']/);
  if (imgMatch) {
    return imgMatch[1];
  }

  return undefined;
}

function parseDate(dateString: string): string {
  try {
    const date = new Date(dateString);
    return date.toISOString().split("T")[0];
  } catch {
    return dateString.substring(0, 10) || "";
  }
}

function slugFromUrl(urlString: string): string {
  const path = urlString.split("?")[0].replace(/\/$/, "");
  return path.split("/").pop() || "";
}

function hashString(str: string, length: number): string {
  return crypto.createHash("md5").update(str).digest("hex").substring(0, length);
}

function entryToPost(itemXml: string): BlogPost | null {
  // Extract link
  const linkMatch = itemXml.match(/<link[^>]*>([^<]+)<\/link>/);
  const urlString = linkMatch ? linkMatch[1].trim() : "";

  if (!urlString) return null;

  const slug = slugFromUrl(urlString) || hashString(urlString, 8);
  const postId = hashString(urlString, 12);

  // Extract title
  const titleMatch = itemXml.match(/<title[^>]*>(?:<!\[CDATA\[(.*?)\]\]>|([^<]+))<\/title>/);
  const title = titleMatch ? (titleMatch[1] || titleMatch[2] || "Untitled") : "Untitled";

  // Extract published date
  const pubDateMatch = itemXml.match(/<pubDate>([^<]+)<\/pubDate>/);
  const date = pubDateMatch ? parseDate(pubDateMatch[1]) : "";

  // Extract description
  const descriptionMatch = itemXml.match(/<description[^>]*>(?:<!\[CDATA\[(.*?)\]\]>|([^<]+))<\/description>/);
  let rawDescription = descriptionMatch ? (descriptionMatch[1] || descriptionMatch[2] || "") : "";
  const description = rawDescription
    .replace(/<[^>]+>/g, "")
    .trim()
    .substring(0, 400);

  // Extract tags/categories
  const tags: string[] = [];
  const categoryMatches = itemXml.matchAll(/<category[^>]*>([^<]*)<\/category>/g);
  for (const match of categoryMatches) {
    if (match[1]) {
      tags.push(match[1]);
    }
  }

  return {
    id: postId,
    title: title.trim(),
    slug: slug,
    url: urlString,
    thumbnail: extractThumbnail(itemXml),
    date: date,
    description: description,
    tags: tags,
  };
}

async function fetchPosts(): Promise<BlogPost[]> {
  try {
    const response = await fetch(MEDIUM_RSS_URL, {
      headers: {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
      },
    });

    if (!response.ok) {
      throw new Error(`Failed to fetch RSS feed: ${response.status}`);
    }

    const rawXml = await response.text();

    // Parse RSS/Atom feed manually
    const posts: BlogPost[] = [];

    // Extract all items
    const itemRegex = /<item[\s\S]*?<\/item>/g;
    const items = rawXml.match(itemRegex) || [];

    items.forEach((itemXml) => {
      const post = entryToPost(itemXml);
      if (post) {
        posts.push(post);
      }
    });

    // Sort by date (newest first)
    posts.sort((a, b) => {
      return new Date(b.date).getTime() - new Date(a.date).getTime();
    });

    return posts;
  } catch (error) {
    console.error("Failed to fetch blog posts:", error);
    return [];
  }
}

async function getCachedPosts(): Promise<BlogPost[]> {
  const now = Date.now() / 1000; // Convert to seconds
  if (now < cache.expiresAt && cache.posts.length > 0) {
    return cache.posts;
  }

  const posts = await fetchPosts();
  cache.posts = posts;
  cache.expiresAt = now + CACHE_TTL;
  return posts;
}

export async function getBlogPosts(): Promise<BlogPost[]> {
  return getCachedPosts();
}

export async function getBlogPost(slug: string): Promise<BlogPost | null> {
  const posts = await getBlogPosts();
  return posts.find((post) => post.slug === slug) || null;
}
